@zugaldia  not sure why circleci is giving this compiler error:

> /home/ubuntu/android-robocar/mobile/app/src/main/java/com/zugaldia/robocar/mobile/MainActivity.java:38: error: MainActivity is not abstract and does not override abstract method onSpeed(RobocarSpeed) in RequestListener
> > Building 74% > :app:compileDebugJavaWithJavacpublic class MainActivity extends AppCompatActivity
> 

 It compiles and works just fine on my side. Any idea?

--- a/mobile/app/src/main/java/com/zugaldia/robocar/mobile/MainActivity.java
+++ b/mobile/app/src/main/java/com/zugaldia/robocar/mobile/MainActivity.java
@@ -24,6 +24,7 @@ import com.zugaldia.robocar.software.webserver.RequestListener;
 import com.zugaldia.robocar.software.webserver.RobocarClient;
 import com.zugaldia.robocar.software.webserver.models.RobocarMove;
 import com.zugaldia.robocar.software.webserver.models.RobocarResponse;
+import com.zugaldia.robocar.software.webserver.models.RobocarSpeed;
 import com.zugaldia.robocar.software.webserver.models.RobocarStatus;
 
 import fi.iki.elonen.NanoHTTPD;
@@ -244,5 +245,11 @@ public class MainActivity extends AppCompatActivity
   @Override
   public RobocarResponse onMove(RobocarMove move) {
     return new RobocarResponse(200, String.format("Moved: %d", move.getKeyCode()));
+  }
+
+  @Override
+  public RobocarResponse onSpeed(RobocarSpeed speed) {
+    return new RobocarResponse(200, String.format(
+        "Speed change: %d/%d", speed.getLeft(), speed.getRight()));
   }
 }

--- a/robocar/app/src/main/java/com/zugaldia/robocar/app/RobocarSpeedChanger.java
+++ b/robocar/app/src/main/java/com/zugaldia/robocar/app/RobocarSpeedChanger.java
@@ -34,8 +34,11 @@ class RobocarSpeedChanger {
 
     public void changeSpeed(RobocarSpeed speed) {
 
-        setLeftSpeed(speed.getLeft());
-        setRightSpeed(speed.getRight());
+        if(speed.getLeft()!=null)
+            setLeftSpeed(speed.getLeft());
+
+        if(speed.getRight()!=null)
+            setRightSpeed(speed.getRight());
     }
 
     public void setLeftSpeed(int speed) {

--- a/robocar/libsoftware/src/main/java/com/zugaldia/robocar/software/webserver/models/RobocarSpeed.java
+++ b/robocar/libsoftware/src/main/java/com/zugaldia/robocar/software/webserver/models/RobocarSpeed.java
@@ -2,22 +2,14 @@ package com.zugaldia.robocar.software.webserver.models;
 
 public class RobocarSpeed {
 
-    private int left;
-    private int right;
+    private Integer left;
+    private Integer right;
 
-    public int getLeft(){
-        return left;
-    }
+    public Integer getLeft(){ return left; }
+    public void setLeft(Integer speed){ left = speed; }
 
-    public void setLeft(int speed){
-        left = speed;
-    }
-
-    public int getRight(){
-        return right;
-    }
-
-    public void setRight(int speed){
+    public Integer getRight(){ return right; }
+    public void setRight(Integer speed){
         right = speed;
     }
 

